### ISSUE TITLE
[Focus] Excluding subtree from focus

### ISSUE THREAD

**Issue by guidezpl** [flutter/flutter#55452]
I'm using an `IndexedStack` with 3 elements and would like to exclude the 2 non-visible subtrees from being focusable. I used to be able to do this but it is no longer working. 

Running the example below, you'll see that even though the 2 other items are marked as non-focusable, focus still goes to their button child.

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  // This widget is the root of your application.
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  int selectedIndex = 0;
  List<int> indices = [0, 1, 2];

Widget build(BuildContext context) {
    debugDumpFocusTree();
    return MaterialApp(
      title: 'focus exclusion repro',
      home: Scaffold(
        body: IndexedStack(
          index: selectedIndex,
          children: [
            for (int index in indices)
              Focus(
                canRequestFocus: index == selectedIndex,
                debugLabel: 'item $index',
                child: FlatButton(
                  child: Text(index.toString()),
                  onPressed: () {},
                ),
              )
          ],
        ),
      ),
    );
  }
}
```

```
[✓] Flutter (Channel master, v1.18.0-6.0.pre, on Mac OS X 10.15.4 19E287, locale en-DE)
    • Flutter version 1.18.0-6.0.pre at /Users/plg/development/flutter
    • Framework revision 84c84fb249 (2 days ago), 2020-04-20 21:35:01 -0400
    • Engine revision 2bde4f0ae4
    • Dart version 2.9.0 (build 2.9.0-1.0.dev a12c36dd97)
```

```
         └─Child 1: FocusScopeNode#ded76(_ModalScopeState<dynamic> Focus Scope [IN FOCUS PATH])
           │ context: FocusScope
           │ IN FOCUS PATH
           │ focusedChildren: FocusNode#c7f9e([PRIMARY FOCUS]),
           │   FocusNode#29a65, FocusNode#562f5(item 0), FocusNode#a1335
           │
           ├─Child 1: FocusNode#562f5(item 0)
           │ │ context: Focus
           │ │
           │ └─Child 1: FocusNode#29a65
           │     context: Focus
           │
           ├─Child 2: FocusNode#b1df1(item 1 [IN FOCUS PATH])
           │ │ context: Focus
           │ │ NOT FOCUSABLE
           │ │ IN FOCUS PATH
           │ │
           │ └─Child 1: FocusNode#c7f9e([PRIMARY FOCUS])
           │     context: Focus
           │     PRIMARY FOCUS <----- Would expect this not to happen because parent is not focusable
           │
           └─Child 3: FocusNode#4fe08(item 2)
             │ context: Focus
             │ NOT FOCUSABLE
             │
             └─Child 1: FocusNode#a1335
                 context: Focus
```

cc @gspencergoog

**Comment by gspencergoog** [comment]
This is (partly) because you're adding a Focus widget (and an associated focus node) and controlling its focusability, but not affecting the one inside of the FlatButton.

Setting `canRequestFocus` on a parent `Focus` node doesn't affect children, by design. Setting `canRequestFocus` to false on a parent `FocusScope`, in contrast, does prevent all the children from being focused. If you can use a FocusScope instead, then this will work.  The problem there is that if you don't want to confine focus traversal to the scope, then this doesn't work for you.

Normally, this would be solved by creating a focusNode and passing it into the FlatButton instead of creating a new Focus widget above it.

The problem that might need to be fixed with the Focus stuff is that when you specify `canRequestFocus` to a `Focus` widget constructor, and also specify a `focusNode` there, the `canRequestFocus` argument overwrites the value that is in the `focusNode`. Since `RawMaterialButton` does pass in a value for `canRequestFocus` corresponding to whether it is enabled or not, this is the case here.  This prevents you from setting the value on the `focusNode` that you want to pass in to widgets that take a `focusNode` to control their focusability from "above" in the widget tree.

I can make it so that the `Focus` widget only sets the value of `canRequestFocus` on the internal node when you don't pass a `focusNode`, but then anyone who sets both `focusNode` and `canRequestFocus` will get their value of `canRequestFocus` ignored and have to set it on the focus node instead.  The other option is to add `canRequestFocus` to all of the widgets that currently take a `focusNode`.

**Comment by guidezpl** [comment]
Thanks, that's a very clear explanation. You're right, FocusScope won't work here because I don't want to confine traversal there.

I worry that either option won't be flexible enough. Example 1: the child widget you want to exclude is from a third-party library. Example 2: you want to exclude a relatively large subtree and don't want to have to manage focus for every focusable descendant. My expectations were driven by the `Semantics` widget's `excludeSemantics` property, as well as the `ExcludeSemantics` widget. Ideally, I would be able to do something like this:

```dart
ExcludeFocus(
   excluding: index == selectedIndex
   child: FlatButton(
     child: Text(index.toString()),
     onPressed: () {},
   ),
),
```

**Comment by gspencergoog** [comment]
OK, check out the solution in #55756, that adds what you're asking for.

**Comment by guidezpl** [comment]
Cool! Added comments

**Comment by guidezpl** [comment]
Marking as closed since https://github.com/flutter/flutter/pull/55756 has landed
